### DISCUSSION TITLE
Understanding  self.control_* usage

### DISCUSSION THREAD

**Discussion by marcelrv** [Louisvdw/dbus-serialbattery#1015] [Q&A] [answered]
Hi,
I've created support for the Seplos v3 BMS. https://github.com/marcelrv/dbus-serialbattery/tree/seplosv3

What I not yet quite understand what is the proper usage / updating of the self.control_* fields:

        # self.control_voltage: float = None
        # self.control_discharge_current: int = None
        # self.control_charge_current: int = None
        # self.control_allow_charge: bool = None
        # self.control_allow_discharge: bool = None

I have populated the voltage & currents with the same data I see in the can-bus (build-in by victron) implementation (in the picture)
            self.control_voltage = spa[0x65] / 100  
            self.control_charge_current = spa[0x66]
            self.control_discharge_current = spa[0x67]
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/5871213/fc7e1e91-2861-4d43-afab-43e987eeffe6)

But what do do with the self.control_allow_charge  & self.control_allow_discharge.
I do have the status of the charge & discharge fets of the BMS. Would this be what I use to put in the allow charge & discharge?

**Comment by mr-manuel** [comment, marked as answer]
You don't need to populate these values. Check the [battery_template.py](https://github.com/Louisvdw/dbus-serialbattery/blob/master/etc/dbus-serialbattery/bms/battery_template.py) to see which values are needed.

The values you mentioned are calculated by the driver itself.
